### DISCUSSION TITLE
Allow static methods to be called off instances

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#6673] [Language Ideas] [answered]
I maintain a code base which has public methods, which used to reference instance values but now are eligible for the `static` keyword.

However, adding that keyword would break all of my consumers, and I can't justify that break.

I was wondering if it would be possible to essentially revoke `CS0176: Static member 'member' cannot be accessed with an instance reference; qualify it with a type name instead`. 

It wouldn't be a breaking change since it currently throws a compiler error, and I'm not sure of any downsides.

**Comment by HaloFour** [comment, marked as answer]
Other than allowing you to soften a breaking change (and it's still a binary breaking change, the method signature is different), I don't see much of a benefit to a change like this.  As for downsides, I think confusing static method calls as instance method calls is a downside enough, which is the reason the language forbids it.  Clarity is important.  To avoid breaking changes the compiler would have to exhaust all potential overloads of an instance member, including with `params` or optional members, before it could even start to consider static members, which would only compound that confusion.

**Reply by RenderMichael** [reply]
This change can be done at lowering, turning this
``` csharp
C c = new C();

c.M();

class C
{
    public static void M() { }
}
```
into
``` csharp
C c = new C();

C.M();

class C
{
    public static void M() { }
}
```
Since the former throws a compiler error, there aren't any breaking changes with this proposal

**Reply by HaloFour** [reply]
I didn't suggest that there would be breaking changes.  I do think it would be confusing:

```cs
C c = new C();

c.M(); // prints instance

class C
{
    public static void M() => Console.WriteLine("static");
    public void M(string s = null) => Console.WriteLine("instance");
}
```

Personally I don't think there's value in blurring the line between static and instance at the call site in such a manner.  That's just my opinion, tho.

**Reply by 333fred** [reply]
> However, adding that keyword would break all of my consumers, and I can't justify that break.

The point Halo was making was that, regardless of whether this was made _source_ compatible (ie, on recompile, no new errors are introduced) it would still be a _binary_ breaking change, as `static void M()` and `void M()` have different signatures in IL. Looking at this [sharplab](https://sharplab.io/#v2:C4LglgNgPgAgTARgLAChUDsCmB3ABAYQAoBKAOgFkESBuVfCuG1VGAZl3gNwG9Vd+O7GABZclEjz4CAvlP5zBHBADYOo8o2I9ZKaUA==) example:

```
        IL_0000: newobj instance void C::.ctor()
        IL_0005: call instance void C::M1()
        IL_000a: call void C::M2()
```

Notice how the il for referencing `M2()` in the call is `void C::M2()`, while the syntax for referencing `M1()` is `instance void C::M1()`. This is a binary break, and if your consumers did not recompile on upgrade their code would break with a `MissingMethodException`.

**Reply by HaloFour** [reply]
Yes, sorry, the breaking change I was referring to was the binary breaking change to the contract of your library.  All consumers would have to be recompiled.  I understand that the purpose of this proposal is to, in theory, make that a simple recompile vs. having to refactor the call sites of these methods to be qualified by the type name.

**Reply by RenderMichael** [reply]
Gotcha, thanks for the explanation, and I understand that this is a non-starter. Just wanted to run it by in a discussion, to see if it would be possible. Thanks!

**Comment by TahirAhmadov** [comment]
I think the solution to your challenge is to 1) create the static methods as suggested but with a slightly different name, 2) mark the instance ones as `[Obsolete("Use static version of this method")]`, 3) and change them to be a pass-through to the `static` counter-parts. Existing consumers will continue working as is; upon recompile, they can modify their code to call the `static` methods.

**Reply by HaloFour** [reply]
C# doesn't permit two methods with the same name and signature where the only difference being that one is an instance method and the other is a static method.

**Reply by TahirAhmadov** [reply]
Yes of course, my bad. I changed it to say with a slightly different name.
